README.rst
Python Client for IAM Policy Troubleshooter API API
===================================================

|stable| |pypi| |versions|

`IAM Policy Troubleshooter API API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-policy-troubleshooter.svg
   :target: https://pypi.org/project/google-cloud-policy-troubleshooter/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-policy-troubleshooter.svg
   :target: https://pypi.org/project/google-cloud-policy-troubleshooter/
.. _IAM Policy Troubleshooter API API: https://cloud.google.com/iam/docs/troubleshooting-access#rest-api/
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/policytroubleshooter/latest
.. _Product Documentation:  https://cloud.google.com/iam/docs/troubleshooting-access#rest-api/

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the IAM Policy Troubleshooter API API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the IAM Policy Troubleshooter API API.:  https://cloud.google.com/iam/docs/troubleshooting-access#rest-api/
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-policy-troubleshooter


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-policy-troubleshooter

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for IAM Policy Troubleshooter API API
   to see other available methods on the client.
-  Read the `IAM Policy Troubleshooter API API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _IAM Policy Troubleshooter API API Product documentation:  https://cloud.google.com/iam/docs/troubleshooting-access#rest-api/
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

### [1.1.2](https://www.github.com/googleapis/python-policy-troubleshooter/compare/v1.1.1...v1.1.2) (2021-07-29)


### Features

* add Samples section to CONTRIBUTING.rst ([#44](https://www.github.com/googleapis/python-policy-troubleshooter/issues/44)) ([a5f4794](https://www.github.com/googleapis/python-policy-troubleshooter/commit/a5f4794092eff9553c847ff65f4e6e1c0d0a5b92))


### Bug Fixes

* enable self signed jwt for grpc ([#48](https://www.github.com/googleapis/python-policy-troubleshooter/issues/48)) ([7186a42](https://www.github.com/googleapis/python-policy-troubleshooter/commit/7186a42e94fd09b663062a09dd3f7baa8906e497))


### Miscellaneous Chores

* release as 1.1.2 ([#49](https://www.github.com/googleapis/python-policy-troubleshooter/issues/49)) ([0247df5](https://www.github.com/googleapis/python-policy-troubleshooter/commit/0247df51703c466a9bcd911696968b8d3533d6bf))

### [1.1.1](https://www.github.com/googleapis/python-policy-troubleshooter/compare/v1.1.0...v1.1.1) (2021-07-21)


### Bug Fixes

* **deps:** pin 'google-{api,cloud}-core', 'google-auth' to allow 2.x versions ([#43](https://www.github.com/googleapis/python-policy-troubleshooter/issues/43)) ([27bf301](https://www.github.com/googleapis/python-policy-troubleshooter/commit/27bf301320a1aad517fd2c9ea42c9b079e4a5cc4))

## [1.1.0](https://www.github.com/googleapis/python-policy-troubleshooter/compare/v1.0.0...v1.1.0) (2021-07-14)


### Features

* add always_use_jwt_access ([#35](https://www.github.com/googleapis/python-policy-troubleshooter/issues/35)) ([0c3e8f7](https://www.github.com/googleapis/python-policy-troubleshooter/commit/0c3e8f7d4554e3d3f515cd6f92c7538482fe6155))


### Bug Fixes

* disable always_use_jwt_access ([f4f810d](https://www.github.com/googleapis/python-policy-troubleshooter/commit/f4f810ddc58b4ebe076457d6d1ba75d023fce572))
* disable always_use_jwt_access ([#39](https://www.github.com/googleapis/python-policy-troubleshooter/issues/39)) ([f4f810d](https://www.github.com/googleapis/python-policy-troubleshooter/commit/f4f810ddc58b4ebe076457d6d1ba75d023fce572))


### Documentation

* omit mention of Python 2.7 in 'CONTRIBUTING.rst' ([#1127](https://www.github.com/googleapis/python-policy-troubleshooter/issues/1127)) ([#30](https://www.github.com/googleapis/python-policy-troubleshooter/issues/30)) ([da0adc2](https://www.github.com/googleapis/python-policy-troubleshooter/commit/da0adc2d0c05be77f06f21e0f34a0d35705832b2))

## [1.0.0](https://www.github.com/googleapis/python-policy-troubleshooter/compare/v0.2.0...v1.0.0) (2021-06-02)


### Features

* bump release level to production/stable ([#22](https://www.github.com/googleapis/python-policy-troubleshooter/issues/22)) ([8411e45](https://www.github.com/googleapis/python-policy-troubleshooter/commit/8411e45cdef6268f51cf54e043ed341303918e41))

## [0.2.0](https://www.github.com/googleapis/python-policy-troubleshooter/compare/v0.1.0...v0.2.0) (2021-05-28)


### Features

* support self-signed JWT flow for service accounts ([ad65aad](https://www.github.com/googleapis/python-policy-troubleshooter/commit/ad65aade357a811bcd41344702d48c497034c9e1))


### Bug Fixes

* add async client to %name_%version/init.py ([ad65aad](https://www.github.com/googleapis/python-policy-troubleshooter/commit/ad65aade357a811bcd41344702d48c497034c9e1))

## 0.1.0 (2021-03-24)


### Features

* generate v1 ([1dd65ce](https://www.github.com/googleapis/python-policy-troubleshooter/commit/1dd65cedfb2ed9c614abe8d3037f4dc31a36a0b8))


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")


nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
]

@nox.session(python=['3.6', '3.7', '3.8', '3.9'])
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'asyncmock', 'pytest-asyncio')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/policytroubleshooter_v1/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python='3.7')
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=['3.6', '3.7'])
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python='3.6')
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx<3.0.0", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )
